File: sdk/java/CaribeFiscal.java
// CaribeFiscal — SDK Java para la API e-CF de la DGII (República Dominicana).
//
// Java 11+ (usa java.net.http.HttpClient). Sin dependencias externas: el cuerpo
// de las peticiones se construye como String JSON; las respuestas se devuelven
// como texto JSON crudo (parsea con tu librería preferida: Jackson/Gson).
//
// Ejemplo:
//   CaribeFiscal cf = new CaribeFiscal("https://ecf.omegaerp.do", "ck_live_xxx");
//   String json = "{ \"ecfType\":\"31\", ... }";
//   String res  = cf.emitInvoice(json);   // -> { "eNCF": "...", ... }

import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpRequest.BodyPublishers;
import java.net.http.HttpResponse;
import java.net.http.HttpResponse.BodyHandlers;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.Map;
import java.util.StringJoiner;

public class CaribeFiscal {

    public static class CaribeFiscalException extends RuntimeException {
        public final int status;
        public final String body;
        public CaribeFiscalException(String message, int status, String body) {
            super(message);
            this.status = status;
            this.body = body;
        }
    }

    private final String baseUrl;
    private final String apiKey;
    private final HttpClient http;

    public CaribeFiscal(String baseUrl, String apiKey) {
        this.baseUrl = baseUrl.replaceAll("/+$", "");
        this.apiKey = apiKey;
        this.http = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(120)).build();
    }

    // ── e-CF ──────────────────────────────────────────────────────────────────
    public String emitInvoice(String jsonBody)            { return req("POST", "/api/v1/ecf/submit", jsonBody); }
    public String getStatus(String trackId)               { return req("GET", "/api/v1/ecf/status/" + enc(trackId), null); }
    public String getInvoice(String encf)                 { return req("GET", "/api/v1/ecf/" + enc(encf), null); }
    public String cancelInvoice(String encf, String motivo){ return req("POST", "/api/v1/ecf/cancel", obj("encf", encf, "motivo", motivo)); }
    public String getQueue()                              { return req("GET", "/api/v1/ecf/queue", null); }
    public String listInvoices(Map<String,String> filters){ return req("GET", "/api/v1/ecf/" + qs(filters), null); }
    public String listInbound(Map<String,String> filters) { return req("GET", "/api/v1/inbound" + qs(filters), null); }

    /** Devuelve los bytes del PDF. */
    public byte[] downloadPdf(String encf) {
        try {
            HttpRequest r = base("/api/v1/ecf/" + enc(encf) + "/pdf").GET().build();
            HttpResponse<byte[]> resp = http.send(r, BodyHandlers.ofByteArray());
            if (resp.statusCode() >= 400) throw new CaribeFiscalException("HTTP " + resp.statusCode(), resp.statusCode(), null);
            return resp.body();
        } catch (IOException | InterruptedException e) {
            throw new CaribeFiscalException("Error de red: " + e.getMessage(), 0, null);
        }
    }

    // ── NCF ───────────────────────────────────────────────────────────────────
    public String getSequences()                          { return req("GET", "/api/v1/ncf/sequences", null); }
    public String nextNcf(String type)                    { return req("GET", "/api/v1/ncf/next/" + enc(type), null); }
    public String configureSequence(String jsonBody)      { return req("POST", "/api/v1/ncf/configure", jsonBody); }
    public String deactivateSequence(String type)         { return req("DELETE", "/api/v1/ncf/" + enc(type), null); }

    // ── RNC ───────────────────────────────────────────────────────────────────
    public String lookupRnc(String rnc)                   { return req("GET", "/api/v1/rnc/" + enc(rnc), null); }
    public String autocompleteRnc(String q)               { return req("GET", "/api/v1/rnc/autocomplete" + qs(Map.of("q", q)), null); }
    public String bulkRnc(String jsonBody)                { return req("POST", "/api/v1/rnc/bulk", jsonBody); }

    // ── Reportes ──────────────────────────────────────────────────────────────
    public String report606(String s, String e)           { return report("606", s, e); }
    public String report607(String s, String e)           { return report("607", s, e); }
    public String report608(String s, String e)           { return report("608", s, e); }
    public String reportIt1(String s, String e)           { return report("it1", s, e); }
    public String reportIsr(String s, String e)           { return report("isr", s, e); }
    public String reportIr2(String jsonBody)              { return req("POST", "/api/v1/reports/ir2", jsonBody); }
    public String reportIr1(String jsonBody)              { return req("POST", "/api/v1/reports/ir1", jsonBody); }
    public String reportTss(String jsonBody)              { return req("POST", "/api/v1/reports/tss", jsonBody); }
    private String report(String tipo, String s, String e){ return req("POST", "/api/v1/reports/" + tipo, obj("startDate", s, "endDate", e)); }

    // ── Empresa ───────────────────────────────────────────────────────────────
    public String getConfig()                             { return req("GET", "/api/v1/company/config", null); }
    public String updateConfig(String jsonBody)           { return req("PUT", "/api/v1/company/config", jsonBody); }

    // ── Núcleo HTTP ─────────────────────────────────────────────────────────
    private HttpRequest.Builder base(String path) {
        return HttpRequest.newBuilder(URI.create(baseUrl + path))
                .timeout(Duration.ofSeconds(120))
                .header("X-API-Key", apiKey)
                .header("Accept", "application/json");
    }

    private String req(String method, String path, String jsonBody) {
        try {
            HttpRequest.Builder b = base(path);
            if (jsonBody != null) {
                b.header("Content-Type", "application/json");
                b.method(method, BodyPublishers.ofString(jsonBody, StandardCharsets.UTF_8));
            } else {
                b.method(method, BodyPublishers.noBody());
            }
            HttpResponse<String> resp = http.send(b.build(), BodyHandlers.ofString());
            if (resp.statusCode() >= 400) {
                throw new CaribeFiscalException("HTTP " + resp.statusCode(), resp.statusCode(), resp.body());
            }
            return resp.body();
        } catch (IOException | InterruptedException e) {
            throw new CaribeFiscalException("Error de red: " + e.getMessage(), 0, null);
        }
    }

    private static String enc(String v) { return URLEncoder.encode(v, StandardCharsets.UTF_8); }

    private static String qs(Map<String, String> params) {
        if (params == null || params.isEmpty()) return "";
        StringJoiner sj = new StringJoiner("&", "?", "");
        params.forEach((k, v) -> { if (v != null && !v.isEmpty()) sj.add(enc(k) + "=" + enc(v)); });
        return sj.toString();
    }

    /** Mini-builder de JSON para cuerpos simples clave/valor (strings). */
    private static String obj(String... kv) {
        StringBuilder sb = new StringBuilder("{");
        for (int i = 0; i + 1 < kv.length; i += 2) {
            if (i > 0) sb.append(',');
            sb.append('"').append(kv[i]).append("\":\"").append(kv[i + 1].replace("\"", "\\\"")).append('"');
        }
        return sb.append('}').toString();
    }
}

File: sdk/rust/caribefiscal.rs
//! CaribeFiscal — SDK Rust para la API e-CF de la DGII (República Dominicana).
//!
//! Dependencias (Cargo.toml):
//!   reqwest    = { version = "0.12", features = ["blocking", "json"] }
//!   serde_json = "1"
//!
//! Ejemplo:
//!   use caribefiscal::CaribeFiscal;
//!   let cf = CaribeFiscal::new("https://ecf.omegaerp.do", "ck_live_xxx");
//!   let res = cf.emit_invoice(&serde_json::json!({ "ecfType": "31", /* ... */ }))?;
//!   println!("{}", res["eNCF"]);

use std::time::Duration;
use reqwest::blocking::Client;
use reqwest::Method;
use serde_json::Value;

#[derive(Debug)]
pub struct CaribeFiscalError {
    pub message: String,
    pub status: u16,
    pub body: Option<Value>,
}

impl std::fmt::Display for CaribeFiscalError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "CaribeFiscalError({}): {}", self.status, self.message)
    }
}
impl std::error::Error for CaribeFiscalError {}

type Result<T> = std::result::Result<T, CaribeFiscalError>;

pub struct CaribeFiscal {
    base_url: String,
    api_key: String,
    client: Client,
}

impl CaribeFiscal {
    pub fn new(base_url: &str, api_key: &str) -> Self {
        let client = Client::builder()
            .timeout(Duration::from_secs(120))
            .build()
            .expect("no se pudo construir el cliente HTTP");
        CaribeFiscal {
            base_url: base_url.trim_end_matches('/').to_string(),
            api_key: api_key.to_string(),
            client,
        }
    }

    // ── e-CF ──────────────────────────────────────────────────────────────────
    pub fn emit_invoice(&self, payload: &Value) -> Result<Value> { self.req(Method::POST, "/api/v1/ecf/submit", Some(payload)) }
    pub fn get_status(&self, track_id: &str) -> Result<Value>    { self.req(Method::GET, &format!("/api/v1/ecf/status/{}", enc(track_id)), None) }
    pub fn get_invoice(&self, encf: &str) -> Result<Value>       { self.req(Method::GET, &format!("/api/v1/ecf/{}", enc(encf)), None) }
    pub fn cancel_invoice(&self, encf: &str, motivo: &str) -> Result<Value> {
        self.req(Method::POST, "/api/v1/ecf/cancel", Some(&serde_json::json!({ "encf": encf, "motivo": motivo })))
    }
    pub fn get_queue(&self) -> Result<Value>                     { self.req(Method::GET, "/api/v1/ecf/queue", None) }
    pub fn list_invoices(&self, query: &[(&str, &str)]) -> Result<Value> { self.req(Method::GET, &format!("/api/v1/ecf/{}", qs(query)), None) }
    pub fn list_inbound(&self, query: &[(&str, &str)]) -> Result<Value>  { self.req(Method::GET, &format!("/api/v1/inbound{}", qs(query)), None) }

    /// Devuelve los bytes del PDF.
    pub fn download_pdf(&self, encf: &str) -> Result<Vec<u8>> {
        let url = format!("{}/api/v1/ecf/{}/pdf", self.base_url, enc(encf));
        let resp = self.client.get(&url).header("X-API-Key", &self.api_key).send()
            .map_err(|e| net_err(&e.to_string()))?;
        let status = resp.status().as_u16();
        let bytes = resp.bytes().map_err(|e| net_err(&e.to_string()))?;
        if status >= 400 {
            return Err(CaribeFiscalError { message: format!("HTTP {}", status), status, body: None });
        }
        Ok(bytes.to_vec())
    }

    // ── NCF ───────────────────────────────────────────────────────────────────
    pub fn get_sequences(&self) -> Result<Value>                { self.req(Method::GET, "/api/v1/ncf/sequences", None) }
    pub fn next_ncf(&self, ecf_type: &str) -> Result<Value>     { self.req(Method::GET, &format!("/api/v1/ncf/next/{}", enc(ecf_type)), None) }
    pub fn configure_sequence(&self, seq: &Value) -> Result<Value> { self.req(Method::POST, "/api/v1/ncf/configure", Some(seq)) }
    pub fn deactivate_sequence(&self, ecf_type: &str) -> Result<Value> { self.req(Method::DELETE, &format!("/api/v1/ncf/{}", enc(ecf_type)), None) }

    // ── RNC ───────────────────────────────────────────────────────────────────
    pub fn lookup_rnc(&self, rnc: &str) -> Result<Value>        { self.req(Method::GET, &format!("/api/v1/rnc/{}", enc(rnc)), None) }
    pub fn autocomplete_rnc(&self, q: &str) -> Result<Value>    { self.req(Method::GET, &format!("/api/v1/rnc/autocomplete{}", qs(&[("q", q)])), None) }
    pub fn bulk_rnc(&self, rncs: &[&str]) -> Result<Value>      { self.req(Method::POST, "/api/v1/rnc/bulk", Some(&serde_json::json!({ "rncs": rncs }))) }

    // ── Reportes ────────────────────────────────────────────────────────────────
    pub fn report_606(&self, s: &str, e: &str) -> Result<Value> { self.report("606", s, e) }
    pub fn report_607(&self, s: &str, e: &str) -> Result<Value> { self.report("607", s, e) }
    pub fn report_608(&self, s: &str, e: &str) -> Result<Value> { self.report("608", s, e) }
    pub fn report_it1(&self, s: &str, e: &str) -> Result<Value> { self.report("it1", s, e) }
    pub fn report_isr(&self, s: &str, e: &str) -> Result<Value> { self.report("isr", s, e) }
    pub fn report_ir2(&self, body: &Value) -> Result<Value>     { self.req(Method::POST, "/api/v1/reports/ir2", Some(body)) }
    pub fn report_ir1(&self, body: &Value) -> Result<Value>     { self.req(Method::POST, "/api/v1/reports/ir1", Some(body)) }
    pub fn report_tss(&self, periodo: &str, empleados: &Value) -> Result<Value> {
        self.req(Method::POST, "/api/v1/reports/tss", Some(&serde_json::json!({ "periodo": periodo, "empleados": empleados })))
    }
    fn report(&self, tipo: &str, start: &str, end: &str) -> Result<Value> {
        self.req(Method::POST, &format!("/api/v1/reports/{}", tipo),
                 Some(&serde_json::json!({ "startDate": start, "endDate": end })))
    }

    // ── Empresa ───────────────────────────────────────────────────────────────
    pub fn get_config(&self) -> Result<Value>                   { self.req(Method::GET, "/api/v1/company/config", None) }
    pub fn update_config(&self, cfg: &Value) -> Result<Value>   { self.req(Method::PUT, "/api/v1/company/config", Some(cfg)) }

    // ── Núcleo HTTP ─────────────────────────────────────────────────────────
    fn req(&self, method: Method, path: &str, body: Option<&Value>) -> Result<Value> {
        let url = format!("{}{}", self.base_url, path);
        let mut rb = self.client.request(method, &url)
            .header("X-API-Key", &self.api_key)
            .header("Accept", "application/json");
        if let Some(b) = body {
            rb = rb.json(b);
        }
        let resp = rb.send().map_err(|e| net_err(&e.to_string()))?;
        let status = resp.status().as_u16();
        let text = resp.text().map_err(|e| net_err(&e.to_string()))?;
        let data: Value = if text.is_empty() { Value::Null } else {
            serde_json::from_str(&text).unwrap_or(Value::Null)
        };
        if status >= 400 {
            let message = data.get("error").and_then(|v| v.as_str())
                .map(|s| s.to_string()).unwrap_or_else(|| format!("HTTP {}", status));
            return Err(CaribeFiscalError { message, status, body: Some(data) });
        }
        Ok(data)
    }
}

fn net_err(msg: &str) -> CaribeFiscalError {
    CaribeFiscalError { message: format!("Error de red: {}", msg), status: 0, body: None }
}

/// Codificación percent simple para segmentos de path / valores de query.
fn enc(s: &str) -> String {
    s.bytes().map(|b| match b {
        b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'_' | b'.' | b'~' => (b as char).to_string(),
        _ => format!("%{:02X}", b),
    }).collect()
}

fn qs(params: &[(&str, &str)]) -> String {
    let parts: Vec<String> = params.iter()
        .filter(|(_, v)| !v.is_empty())
        .map(|(k, v)| format!("{}={}", enc(k), enc(v)))
        .collect();
    if parts.is_empty() { String::new() } else { format!("?{}", parts.join("&")) }
}
